src/tabpfn/inference.py

```diff
@@ -808,3 +790,73 @@ def _move_and_squeeze_output(
     if isinstance(output, dict):
         return {k: v.to(device) for k, v in output.items()}
     return output.squeeze(1).to(device)
+
+
+class _PerDeviceModelCache:
+    """Maintains a copy of a model on each device."""
+
+    def __init__(self, model: Architecture) -> None:
+        super().__init__()
+        self._model = model
+        self._on_device_cache: dict[torch.device, Architecture] = {}
+        self._on_device_cache_lock = Lock()
+
+    def get(self, device: torch.device, *, multiple_devices: bool) -> Architecture:
+        """Return the model on the specified device.
+
+        Return the model from the cache, if present, otherwise copy the model to the
+        device.
+
+        Args:
+            device: The device to get the model for.
+            multiple_devices:
+                If True, indicates that the model is being used on multiple devices.
+                Thus, the model is deepcopied before being moved to the target device,
+                allowing other copies of the model to be on other devices.
+                If False, then the model is moved to the target device without copying
+                to save time.
+        """
+        with self._on_device_cache_lock:
+            not_on_device = device not in self._on_device_cache
+
+            if not_on_device:
+                if multiple_devices:
+                    self._on_device_cache[device] = deepcopy(self._model)
+                else:
+                    self._on_device_cache.clear()
+                    self._on_device_cache[device] = self._model
```

Comment:
```suggestion
                    self._on_device_cache[device] = deepcopy(self._model) if multiple_devices else self._model
```

Then we can remove the if/else which I think is harder to read.